src/raftstore/store/worker/apply.rs

```diff
@@ -2029,180 +2247,332 @@ impl Runner {
+            match self.region_task_senders.get(&region_proposal.region_id) {
+                Some(tx) => {
+                    tx.unbounded_send(RegionTask::proposal(region_proposal))
+                        .unwrap_or_else(|e| {
+                            removed = true;
+                            let task = e.into_inner();
+                            if let RegionTask::Proposal(p) = task {
+                                stale_proposal = Some(p);
+                            } else {
+                                panic!("Proposal task is expected.")
+                            }
+                        });
+                }
                 None => {
-                    error!("[region {}] is missing", apply.region_id);
-                    continue;
+                    removed = true;
+                    stale_proposal = Some(region_proposal);
                 }
-                Some(e) => e.take().unwrap(),
-            };
-            delegate.metrics = ApplyMetrics::default();
-            delegate.term = apply.term;
-
-            {
-                let mut ctx = ApplyContext::new(&mut core, &mut self.delegates);
-                delegate.handle_raft_committed_entries(&mut ctx, apply.entries);
             }
-
-            if delegate.pending_remove {
-                delegate.destroy();
-                self.delegates.remove(&apply.region_id);
-            } else {
-                *self.delegates.get_mut(&apply.region_id).unwrap() = Some(delegate);
+            if removed {
+                let stale_proposal = stale_proposal.unwrap();
+                for p in stale_proposal.props {
+                    let cmd = PendingCmd::new(p.index, p.term, p.cb);
+                    notify_region_removed(stale_proposal.region_id, stale_proposal.id, cmd);
+                }
+                self.region_task_senders.remove(&stale_proposal.region_id);
             }
         }
 
-        // Write to engine
-        // raftsotre.sync-log = true means we need prevent data loss when power failure.
-        // take raft log gc for example, we write kv WAL first, then write raft WAL,
-        // if power failure happen, raft WAL may synced to disk, but kv WAL may not.
-        // so we use sync-log flag here.
-        core.write_to_db(&self.db);
-
-        for region_id in core.merged_regions.drain(..) {
-            if let Some(mut e) = self.delegates.remove(&region_id) {
-                e.as_mut().unwrap().destroy();
-            }
-        }
+        APPLY_PROPOSAL.observe(propose_num as f64);
+    }
 
-        if !core.apply_res.is_empty() {
-            self.notifier.send(TaskRes::Applys(core.apply_res)).unwrap();
-        }
+    fn spawn_region(
+        apply_pool: &FuturePool<PoolContext>,
+        delegate: ApplyDelegate,
+        rx: UnboundedReceiver<RegionTask>,
+    ) {
+        apply_pool
+            .spawn(move |_| {
+                rx.fold(delegate, move |mut delegate, task| {
```

Comment:
I prefer `loop_fn` to `fold`.

Reply:
can you define a closure function outside? The intention hierarchy is too deep below.